src/components/App.tsx

```diff
@@ -1,6 +1,121 @@
-export const App = () => {
-  return (
-    <div>
-    </div>
-  );
+import { Component, Fragment } from 'react';
+import debounce from 'lodash/debounce';
+import { Input, Dropdown, Weather } from "./common";
+import { TWeather, Units } from "./types/types";
+import css from './styles.module.css';
+
+interface AppState {
+  city: string;
+  unit: Units;
+  newCityInput: string;
+
+  weatherInfo: TWeather | null;
+  locationErr: boolean;
+  weatherErr: boolean;
+  isLocationLoading: boolean;
+  isWeatherLoading: boolean;
+  error: string;
+}
+
+export class App extends Component<{}, AppState> {
+  state: AppState = {
+    city: '',
+    unit: 'metric',
+    newCityInput: '',
+    weatherInfo: null,
+    locationErr: false,
+    weatherErr: false,
+    isLocationLoading: true,
+    isWeatherLoading: true,
+    error: '',
+  };
+
+  localStorageKey = 'city';
+
+  debouncedSearch = debounce((newCityInput: AppState['newCityInput']) => {
+    this.getWeatherData(newCityInput || this.state.city, this.state.unit);
+  }, 1500);
+
+  searchHandler = (newCityInput: AppState['newCityInput']) => {
+    this.setState({ newCityInput, isWeatherLoading: true, weatherErr: false, error: '' });
+    this.debouncedSearch(newCityInput);
+  }
+
+  toggleUnitsHandler =  (unit: AppState['unit']) => {
+    this.setState({ unit })
+    this.getWeatherData(this.state.newCityInput || this.state.city, unit);
+  };
+
+  getWeatherData = (city: AppState['city'], unit: AppState['unit']) => {
+    let urlParams = new URLSearchParams({ q: city, appid: `${process.env.REACT_APP_APPID}`, units: unit }).toString();
+    fetch(`https://api.openweathermap.org/data/2.5/weather?${urlParams}`)
+      .then(weather => {
+        if (weather.ok) {
+          return weather.json()
+        }  else {
+          console.log(weather);
+          if (weather.status === 404) {
+            this.setState({ error: '404'});
+          }
+          throw new Error('Network err')
+        }})
+      .then(weatherInfo => {
+        this.setState( { weatherInfo, isWeatherLoading: false, error: ''} )
+      })
+      .catch(() => {
+        this.setState( { weatherErr: true, isWeatherLoading: false, weatherInfo: null})
+      })
+  }
+
+  componentDidMount() {
+    const city = localStorage.getItem(this.localStorageKey) ?? '[]';
+    if (city.length) {
+      this.setState({ city });
+    }
+    if (this.state.city === '' && this.state.isLocationLoading){
+      fetch('http://ip-api.com/json/')
+        .then(location => {
+          if (location.ok) {
+            return location.json()
+          } else {
+            return new Error('Network error');
```

Comment:
нужно кидать ошибку, чтобы отловить ее в .catch
```suggestion
            throw new Error('Network error');
```